lib/Conversion/TritonGPUToLLVM/Utility.cpp

```diff
@@ -348,18 +349,18 @@ static Value commonShflSync(Location loc, ConversionPatternRewriter &rewriter,
       (*shfl)(dOpr, aOpr, maskOpr);
     }
     break;
-  case NVVM::ShflKind::up:
-    //assert(shuffleType == "up" && "Only shfl_bfly and shfl_up are supported");
+  case NVVM::ShflKind::up: {
     Value mask = icmp_slt(laneId, i);
     Value delta = sub(laneId, i);
     Value index = select(mask, laneId, delta);
-    Value byteOffset = i32_val(2);
-    Value permuteAddr = shl(index, byteOffset);
-    auto shfl = builder.create("ds_bpermute_b32");
-    auto dOpr = builder.newOperand("=v");
-    auto addrOpr = builder.newOperand(permuteAddr, "v");
-    auto aOpr = builder.newOperand(val, "v");
-    (*shfl)(dOpr, addrOpr, aOpr);
+    permute(index, "ds_bpermute_b32");
+    break;
+  }
+  case NVVM::ShflKind::idx:
+    permute(i, "ds_bpermute_b32");
```

Comment:
so shfl.idx means each workitem getting the value from an arbitrary workitem in the same wavefront?

Reply:
> so shfl.idx means each workitem getting the value from an arbitrary workitem in the same wavefront?
yes, that is right.